File: app/src/main/java/org/weijian/mydriversrss/activity/DisplayActivity.java
package org.weijian.mydriversrss.activity;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import org.jetbrains.annotations.NotNull;
import org.weijian.mydriversrss.EndlessOnScrollListener;
import org.weijian.mydriversrss.HttpMethods;
import org.weijian.mydriversrss.ImageManager;
import org.weijian.mydriversrss.News;
import org.weijian.mydriversrss.NewsEntity;
import org.weijian.mydriversrss.R;
import org.weijian.mydriversrss.request.NewsRequest;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeSet;

import butterknife.BindView;
import butterknife.ButterKnife;
import rx.Observer;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static org.weijian.mydriversrss.Constants.NEWS_ITEM_TYPE_IMAGE;
import static org.weijian.mydriversrss.Constants.NEWS_ITEM_TYPE_MULTI_IMAGES;
import static org.weijian.mydriversrss.Constants.NEWS_ITEM_TYPE_NO_IMAGE;

/**
 * Created by weijian on 5/3/16.
 */
public class DisplayActivity extends AppCompatActivity implements SwipeRefreshLayout.OnRefreshListener {
    @BindView(R.id.toolbar) Toolbar mToolbar;
    @BindView(R.id.recycler_view) RecyclerView mRecyclerView;
    @BindView(R.id.progress_bar) ProgressBar mProgressBar;
    @BindView(R.id.swipe_refresh) SwipeRefreshLayout mSwipeRefreshLayout;

    private HttpMethods mHttpMethods;
    private NewsAdapter mAdapter;
    private Map<ImageView, String> map = new HashMap<>();

    @Override
    public void onRefresh() {
        retrieveNews("0");
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.menu_display, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_refresh:
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_display);

        ButterKnife.bind(this);

        setSupportActionBar(mToolbar);

        mAdapter = new NewsAdapter(getBaseContext());
        mAdapter.setHasStableIds(true);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getBaseContext()));
        mRecyclerView.setItemAnimator(new DefaultItemAnimator());
        mRecyclerView.addOnScrollListener(
                new EndlessOnScrollListener() {
                    @Override
                    public void onScrolledToEnd(@NotNull RecyclerView recyclerView) {
                        RecyclerView.Adapter adapter = recyclerView.getAdapter();
                        if (adapter instanceof NewsAdapter) {
                            if (((NewsAdapter) adapter).newsSet.size() != 0) {
                                int articleId = ((NewsAdapter) adapter).newsSet.last().getArticleId();
                                retrieveNews(String.valueOf(articleId));
                            }
                        }
                    }
                }
        );
        mRecyclerView.setAdapter(mAdapter);

        mSwipeRefreshLayout.setOnRefreshListener(this);
        mHttpMethods = HttpMethods.getInstance(getApplicationContext());
        ImageManager.getInstance(getApplicationContext()).setOnImageLoadedListener(listener);
        retrieveNews("0");
    }

    Observer<NewsEntity> observer = new Observer<NewsEntity>() {
        @Override
        public void onCompleted() {
            if (mProgressBar.getVisibility() == View.VISIBLE) {
                mProgressBar.setVisibility(View.GONE);
            }
            if (mSwipeRefreshLayout.isRefreshing()) {
                mSwipeRefreshLayout.setRefreshing(false);
            }
        }

        @Override
        public void onError(Throwable e) {

        }

        @Override
        public void onNext(NewsEntity newsEntity) {
            mAdapter.addNewsCollections(newsEntity.getData().getNews());
        }
    };

    private void retrieveNews(String minId) {
        mHttpMethods.getRetrofit("http://dt.kkeji.com", true, true)
                .create(NewsRequest.class)
                .getNews("411203410", "ebbfb000", "39223325", minId, "1", "0")
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer);
    }

    @Override
    protected void onStop() {
        super.onStop();
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    private ImageManager.onImageLoaded listener = new ImageManager.onImageLoaded() {
        @Override
        public void onFinish(ImageView view, Bitmap bitmap, String imageUrl) {
            String url = map.get(view);
            if (url != null && url.equals(imageUrl)) {
                view.setImageBitmap(bitmap);
            }
        }
    };

    class NewsAdapter extends RecyclerView.Adapter {
        private Context mContext;

        public NewsAdapter(Context context) {
            mContext = context;
        }

        class ItemViewHolder extends RecyclerView.ViewHolder {
            public @BindView(R.id.news_title) TextView titleView;
            public @BindView(R.id.news_pub_date) TextView dateView;
            public ImageView[] imageViews;

            public ItemViewHolder(View itemView, int viewType) {
                super(itemView);
                ButterKnife.bind(this, itemView);
                if (viewType == NEWS_ITEM_TYPE_MULTI_IMAGES)
                    imageViews = new ImageView[]{
                            (ImageView) itemView.findViewById(R.id.news_image1),
                            (ImageView) itemView.findViewById(R.id.news_image2),
                            (ImageView) itemView.findViewById(R.id.news_image3)
                    };
                else {
                    imageViews = new ImageView[]{
                            (ImageView) itemView.findViewById(R.id.news_image1)
                    };
                }
            }
        }

        private TreeSet<News> newsSet = new TreeSet<>();
        private News[] newsArray;


        public void addNewsCollections(Collection<News> newsCollection) {
            int size = newsSet.size();
            newsSet.addAll(newsCollection);
            notifyArray(size);
        }

        private void notifyArray(int size) {
            if (size != newsSet.size()) {
                newsArray = newsSet.toArray(new News[newsSet.size()]);
                notifyDataSetChanged();
            }
        }

        public void addNews(News news) {
            int size = newsSet.size();
            newsSet.add(news);
            notifyArray(size);
        }


        @Override
        public int getItemViewType(int position) {
            return newsArray[position].getImgs().size();
        }

        @Override
        public long getItemId(int position) {
            return newsArray[position].getArticleId();
        }

        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            int layoutId = 0;
            switch (viewType) {
                case NEWS_ITEM_TYPE_NO_IMAGE:
                    layoutId = R.layout.news_item;
                    break;
                case NEWS_ITEM_TYPE_IMAGE:
                    layoutId = R.layout.news_item_image;
                    break;
                case NEWS_ITEM_TYPE_MULTI_IMAGES:
                    layoutId = R.layout.news_item_multi_images;
                    break;
            }
            View view = LayoutInflater.from(mContext).inflate(layoutId, parent, false);
            return new ItemViewHolder(view, viewType);
        }


        public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
            News news = newsArray[position];
            if (holder instanceof ItemViewHolder) {
                ((ItemViewHolder) holder).titleView.setText(news.getTitle());
                for (int i = 0; i < news.getImgs().size(); i++) {
                    ImageView imageView = ((ItemViewHolder) holder).imageViews[i];
                    String imageUrl = news.getImgs().get(i);
                    map.put(imageView, imageUrl);
                    Bitmap bitmap = ImageManager.getInstance(mContext).getBitmap(imageUrl, imageView, 0, 0);
                    imageView.setImageBitmap(bitmap);

                }
            }

        }

        @Override
        public int getItemCount() {
            return newsSet.size();
        }
    }

}

File: app/src/main/java/org/weijian/mydriversrss/DisplayActivity.kt
package org.weijian.mydriversrss

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.support.design.widget.AppBarLayout
import android.support.v4.content.LocalBroadcastManager
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.*
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import java.text.SimpleDateFormat
import java.util.*

class DisplayActivity : AppCompatActivity(), SwipeRefreshLayout.OnRefreshListener {

    private lateinit var mToolbar: Toolbar
    private lateinit var mRecyclerView: RecyclerView
    private lateinit var mStatusReciver: StatusReceiver
    private lateinit var mProgressBar: ProgressBar
    private lateinit var mSwipeRefreshLayout: SwipeRefreshLayout
    private lateinit var mAppbarLayout: AppBarLayout
    private lateinit var mStatusIntentFilter: IntentFilter


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_display)

        // Init variables
        mToolbar = findViewById(R.id.toolbar) as Toolbar
        mRecyclerView = findViewById(R.id.recycler_view) as RecyclerView
        mProgressBar = findViewById(R.id.progress_bar) as ProgressBar
        mSwipeRefreshLayout = findViewById(R.id.swipe_refresh) as SwipeRefreshLayout
        mAppbarLayout = findViewById(R.id.appbarLayout) as AppBarLayout


        // get or generate rss parameters
        val signId = savedInstanceState?.getString(Constants.RSS_SIGN_ID) ?: { Random().nextInt().toString() }()
        val xaId = savedInstanceState?.getString(Constants.RSS_XAID) ?: {
            // Generate random XAID
            var stringBuilder = StringBuilder()
            val random = Random()
            kotlin.repeat(8) {
                stringBuilder.append(Integer.toHexString(random.nextInt(16)))
            }
            stringBuilder.toString()
        }()
        val udId = savedInstanceState?.getString(Constants.RSS_UDID) ?: { Random().nextLong().toString() }()
        val minId = savedInstanceState?.getString(Constants.RSS_MINID) ?: "0"

        // Set toolbar
        setSupportActionBar(mToolbar)
        // Set RecyclerView layoutManager
        var linearLayoutManager = LinearLayoutManager(this.baseContext)
        mRecyclerView.layoutManager = linearLayoutManager
        mRecyclerView.itemAnimator = DefaultItemAnimator()
        mRecyclerView.addOnScrollListener(object : EndlessOnScrollListener() {
            override fun onScrolledToEnd(recyclerView: RecyclerView) {
                val adapter = recyclerView.adapter as NewsAdapter
                adapter.getMore()
                adapter.isGetting = true
            }
        })
        val adapter = NewsAdapter(this.baseContext, signId = signId, xaId = xaId, minId = minId, udId = udId)
        adapter.setHasStableIds(true)
        mRecyclerView.adapter = adapter

        // Set RefreshLayout Listener
        mSwipeRefreshLayout.setOnRefreshListener(this)

        // register broadcast receiver
        mStatusIntentFilter = IntentFilter(Constants.BROADCAST_ACTION)
        mStatusIntentFilter.addCategory(Intent.CATEGORY_DEFAULT)
        mStatusReciver = StatusReceiver()
        LocalBroadcastManager.getInstance(this).registerReceiver(mStatusReciver, mStatusIntentFilter)

        // TODO delete this after test
        mProgressBar.visibility = View.GONE

    }


    override fun onStop() {
        super.onStop()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mStatusReciver)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.menu_display, menu)
        return true
    }

    override fun onRefresh() {
        val adapter = mRecyclerView.adapter as NewsAdapter
        adapter.refresh()
    }


    inner class StatusReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val status = intent.extras.get(Constants.EXTENDED_DATA_STATUS)
            when (status) {
                Constants.STATE_ACTION_COMPLETE -> {
                    Log.d("BROADCAST_RECEIVER", "STATE_ACTION_COMPLETE")
                    if (mSwipeRefreshLayout.isRefreshing) {
                        mSwipeRefreshLayout.isRefreshing = false
                    }
                }
            }
        }

    }

    inner class NewsAdapter constructor(val context: Context, var signId: String, var xaId: String
                                        , var minId: String, var udId: String) :
            RecyclerView.Adapter<RecyclerView.ViewHolder>(), NewsResultReceiver.NewsReceiver {
        private lateinit var mServiceIntent: Intent

        var newsList: MutableList<News> = mutableListOf()
        var isGetting: Boolean = false
            get set
        private var mResultReceiver: NewsResultReceiver

        init {
            mResultReceiver = NewsResultReceiver(Handler(), receiver = this)
            mServiceIntent = Intent(context, RssPullService::class.java)
            mServiceIntent.putExtra(Constants.RSS_SIGN_ID, signId)
            mServiceIntent.putExtra(Constants.RSS_XAID, xaId)
            mServiceIntent.putExtra(Constants.RSS_UDID, udId)
            mServiceIntent.putExtra(NewsResultReceiver.NEWS_RESULT_RECEIVER, mResultReceiver)
            refresh()
        }

        fun refresh() {
            // start RssPullService at startup
            mServiceIntent.putExtra(Constants.RSS_MINID, 0)
            mServiceIntent.putExtra(Constants.RECYCLER_ACTION, Constants.RECYCLER_ACTION_REFRESH)
            startService(mServiceIntent)
        }

        fun getMore() {
            if (!isGetting) {
                minId = newsList.last().articleId.toString()
                mServiceIntent.putExtra(Constants.RSS_MINID, minId)
                mServiceIntent.putExtra(Constants.RECYCLER_ACTION, Constants.RECYCLER_ACTION_GET)
                startService(mServiceIntent)
            }
        }

        /**
         *  Implements NewsResultReceiver.Receiver
         * */
        override fun onReceiveResult(resultCode: Int, resultData: Bundle?) {
            when (resultCode) {
                Constants.STATE_FETCH_COMPLETE -> {
                    if (resultData != null) {
                        val action = resultData.getInt(Constants.RECYCLER_ACTION)
                        // Refresh
                        val newsArray = resultData.getSerializable(Constants.NEWS_ITEM_RESULT) as Array<News>
                        if (action == Constants.RECYCLER_ACTION_REFRESH) {
                            val new = mutableSetOf(*newsArray)
                            new.removeAll(newsList)
                            val changeList = new.toMutableList()
                            Collections.sort(changeList) { lhs, rhs ->
                                when {
                                    lhs.pubTime > rhs.pubTime -> 1
                                    lhs.pubTime < rhs.pubTime -> -1
                                    else -> 0
                                }
                            }
                            for (item in changeList) {
                                newsList.add(0, item)
                            }
                            if (changeList.size > 0)
                                notifyItemRangeInserted(0, changeList.size)
                        } else if (action == Constants.RECYCLER_ACTION_GET) {
                            //Get more data
                            val size = newsList.size
                            newsList.addAll(newsArray)
                            notifyItemRangeInserted(size - 1, newsList.size - size)
                            isGetting = false
                        }
                    }
                }
            }
        }

        override fun getItemCount(): Int {
            return newsList.size
        }

        override fun getItemViewType(position: Int): Int {
            return newsList[position].images.size
        }

        override fun getItemId(position: Int): Long {
            return newsList[position].articleId.toLong()
        }

        override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
            var newsHolder = viewHolder as NewsHolder
            val newsItem = newsList[position]
            val itemType = newsItem.images.size
            //TODO Add handler when itemType is not 0
            if (newsHolder.imageViews != null) {
                for (i in 0..newsHolder.imageViews!!.size - 1) {
                    newsHolder.imageViews?.get(i)?.imageUrl = newsItem.images[i]
                }
            }
            newsHolder.titleView.text = newsItem.title
            newsHolder.titleView.paint?.isFakeBoldText = true
            newsHolder.titleView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 16F)
            val dateFormat = SimpleDateFormat("MM-dd HH:mm")
            newsHolder.pubDateView.text = dateFormat.format(Date(newsItem.pubTime * 1000))
        }

        override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder? {
            val layoutId = when (viewType) {
                Constants.NEWS_ITEM_TYPE_NO_IMAGE -> R.layout.news_item
                Constants.NEWS_ITEM_TYPE_IMAGE -> R.layout.news_item_image
                Constants.NEWS_ITEM_TYPE_MULTI_IMAGES -> R.layout.news_item_multi_images
                else -> 0
            }
            var newsCard = LayoutInflater.from(context).inflate(layoutId, parent, false) as CardView
            var newsHolder = NewsHolder(newsCard, viewType)
            return newsHolder
        }

        inner class NewsHolder constructor(var cardView: CardView, viewType: Int) : RecyclerView.ViewHolder(cardView) {
            lateinit var titleView: TextView
            lateinit var commentCountView: TextView
            lateinit var pubDateView: TextView
            var imageViews: Array<AsyncImageView>? = null

            init {
                titleView = cardView.findViewById(R.id.news_title) as TextView
                pubDateView = cardView.findViewById(R.id.news_pub_date) as TextView
                commentCountView = cardView.findViewById(R.id.news_comments_count) as TextView
                //TODO Add handler when itemType is not 0
                when (viewType) {
                    Constants.NEWS_ITEM_TYPE_IMAGE -> {
                        imageViews = arrayOf(cardView.findViewById(R.id.news_image) as AsyncImageView)
                    }
                    Constants.NEWS_ITEM_TYPE_MULTI_IMAGES -> {
                        imageViews = arrayOf(
                                cardView.findViewById(R.id.news_image1) as AsyncImageView,
                                cardView.findViewById(R.id.news_image2) as AsyncImageView,
                                cardView.findViewById(R.id.news_image3) as AsyncImageView
                        )
                    }
                }
            }
        }
    }
}
